graph/xml_parser.py

```diff
@@ -204,23 +208,108 @@ def _operator_as_child(self, value, scan):
+                        scan,
+                        child['extend_definition'],
+                        child['negate'],
+                        child['comment']))
             elif 'value_id' in child:
                 out['node'].append(child)
             else:
                 raise ValueError('error - unknown child')
         return out
 
-    def _find_definition_by_id(self, scan, id, negate_status):
+    def _find_definition_by_id(self, scan, id, negate_status, comment):
         for definition in scan['definitions']:
             if definition['id'] == id:
                 definition['node'][0]['negate'] = negate_status
+                definition['node'][0]['comment'] = comment
                 return self._operator_as_child(definition['node'][0], scan)
+
+    def deeper_in_criteria_comments(self, criteria):
+        if criteria.get('operator') is None:
+            comments = dict(
+                operator='AND',
+                comment=criteria.get('comment'),
+                node=[])
+        else:
+            comments = dict(
+                operator=criteria.get('operator'),
+                comment=criteria.get('comment'),
+                node=[])
+        for criterion in criteria:
+            if criterion.get('operator') is not None:
+                comments['node'].append(
+                    self.deeper_in_criteria_comments(criterion))
+            else:
+                if criterion.get('definition_ref') is not None:
+                    comments['node'].append(
+                        dict(
+                            extend_definition=criterion.get('definition_ref'),
+                            comment=criterion.get('comment')))
+                else:
+                    comments['node'].append(
+                        dict(
+                            value_id=criterion.get('test_ref'),
+                            comment=criterion.get('comment')))
+        return comments
+
+    def prepare_definition_comments(self, oval_definitions):
+        ns = {
+            'oval-definitions': 'http://oval.mitre.org/XMLSchema/oval-definitions-5'
+        }
+        definitions = []
+        for definition in oval_definitions:
+            comment_definition = dict(id=definition.get('id'), node=[])
+            criteria = definition.find('.//oval-definitions:criteria', ns)
+            comment_definition['node'].append(
+                self.deeper_in_criteria_comments(criteria))
+            definitions.append(comment_definition)
+        return definitions
+
+    def is_definition_in_array(self, definition_, array):
+        for definition in array:
+            if definition_['id'] == definition['id']:
+                return True
+        return False
+
+    def help_fill_comments(self, comments, nodes):
+        out = nodes
+        if 'operator' in out:
+            out['comment'] = comments['comment']
+            for i in range(len(out['node'])):
```

Comment:
The loop contains occurences of `out['node'][i]` and `comments['node'][i]`, check out how to use `enumerate` and `zip` in `for` loops to make this more Pythonic with less square brackets.